src/QsCompiler/Transformations/ClassicallyControlled.cs

```diff
@@ -289,25 +266,26 @@ private TypeArgsResolution GetCombinedTypeResolution(TypeArgsResolution outer, T
                         && !TypedExpression.IsPartialApplication(expr.Item.Expression)
                         && call.Item1.Expression is ExpressionKind.Identifier)
                     {
-                        // We are dissolving the application of arguments here, so the call's type argument
-                        // resolutions have to be moved to the 'identifier' sub expression.
-
-                        var callTypeArguments = expr.Item.TypeArguments;
-                        var idTypeArguments = call.Item1.TypeArguments;
-                        var combinedTypeArguments = this.GetCombinedTypeResolution(callTypeArguments, idTypeArguments);
-
-                        // This relies on global callables being the only things that have type parameters.
                         var newCallIdentifier = call.Item1;
-                        if (combinedTypeArguments.Any()
-                            && newCallIdentifier.Expression is ExpressionKind.Identifier id
-                            && id.Item1 is Identifier.GlobalCallable global)
+                        var callTypeArguments = expr.Item.TypeParameterResolutions;
+
+                        // This relies on anything having type parameters must be a global callable.
+                        if (newCallIdentifier.Expression is ExpressionKind.Identifier id
+                            && id.Item1 is Identifier.GlobalCallable global
+                            && callTypeArguments.Any())
                         {
+                            // We are dissolving the application of arguments here, so the call's type argument
+                            // resolutions have to be moved to the 'identifier' sub expression.
+                            var combination = new TypeResolutionCombination(expr.Item);
+                            var combinedTypeArguments = combination.CombinedResolutionDictionary.Where(kvp => kvp.Key.Item1.Equals(global.Item)).ToImmutableDictionary();
```

Comment:
How come the keys need to be filtered here? I.e. why isn't the CombinedResolutionsDictionary guaranteeing that all keys belong to the same parent? And if they are from the same parent, it would actually be nice to have a handle in the resolution that specifies that parent.

Reply:
I see the filtering done here as something that is situation-dependent and not something that fits into the narrow definition of combining type parameter resolutions. It seems better to me to return all the combinations, and if the caller wants to filter those by anything, they can do that themselves, rather than assuming the user only wants part of the data and filtering out the other parameters in the tool.

Reply:
Another thought: If we find later that this specific kind of filtering is done quite often after getting the results from the tool, we can always add some functionality to it to do that automatically. We can always add onto the public-facing content. But if we add it now and it doesn't come up often, we can't remove it without a breaking change, so I would prefer to be conservative with regard to this until we can see lots of use cases that use the filter.

Reply:
Either one is ok for me; just as a question to think about: What would be a valid expression where the resolutions dictionary contains keys for more than one parent?

Reply:
Thanks Bettina! It'll be good to keep in mind what kinds of varieties of inputs this tool can handle going forward. I will keep this in mind.